### ISSUE TITLE
Deepcopy for program is unnecessary 

### ISSUE THREAD

**Issue by wulfihm** [trevorstephens/gplearn#136]
Lines:
https://github.com/trevorstephens/gplearn/blob/208e4f965ee53b21788167f2c47c176fe02f1cf3/gplearn/_program.py#L518
https://github.com/trevorstephens/gplearn/blob/208e4f965ee53b21788167f2c47c176fe02f1cf3/gplearn/_program.py#L628

We do not need a deepcopy.
The only objects stored in the self.program are the math functions via the wrapper class from the functions.py which might make us think we need a deepcopy. But we do not change anything in these objects. What we do is that we delete or replace these objects with other objects, but for that we do not need a deepcopy and a shallow copy is sufficient.

The reason to replace the deepcopy with a shallow copy is that it produces an overhead that is unnecessary. I noticed that when I ran line_profiler on the point_mutation.
Beside, the array slicing in the other genetic operators creates shallow copies as well.

**With deepcopy:**
Time is in 1e-06s
```

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   656                                               @profile
   657                                               def point_mutation(self, random_state):
   658                                                   """Perform the point mutation operation on the program.
   659                                           
   660                                                   Point mutation selects random nodes from the embedded program to be
   661                                                   replaced. Terminals are replaced by other terminals and functions are
   662                                                   replaced by other functions that require the same number of arguments
   663                                                   as the original node. The resulting tree forms an offspring.
   664                                           
   665                                                   Parameters
   666                                                   ----------
   667                                                   random_state : RandomState instance
   668                                                       The random number generator.
   669                                           
   670                                                   Returns
   671                                                   -------
   672                                                   program : list
   673                                                       The flattened tree representation of the program.
   674                                           
   675                                                   """
   676      1165   15487576.0  13294.1     99.5          program = deepcopy(self.program)
   677                                           
   678                                                   # Get the nodes to modify
   679      1165       2856.0      2.5      0.0          mutate = np.where([True if (random_state.uniform() <
   680                                                                               self.p_point_replace)
   681                                                                      else False
   682      1165      48160.0     41.3      0.3                             for _ in range(len(program))])[0]
   683                                           
   684      2097       6439.0      3.1      0.0          for node in mutate:
   685       932       1202.0      1.3      0.0              if isinstance(program[node], _Function):
   686       413        438.0      1.1      0.0                  arity = program[node].arity
   687                                                           # Find a valid replacement with same arity
   688       413        488.0      1.2      0.0                  replacement = len(self.arities[arity])
   689       413       3963.0      9.6      0.0                  replacement = random_state.randint(replacement)
   690       413        399.0      1.0      0.0                  replacement = self.arities[arity][replacement]
   691       413        415.0      1.0      0.0                  program[node] = replacement
   692                                                       else:
   693                                                           # We've got a terminal, add a const or variable
   694       519        431.0      0.8      0.0                  if self.const_range is not None:
   695       519       4317.0      8.3      0.0                      terminal = random_state.randint(self.n_features + 1)
   696                                                           else:
   697                                                               terminal = random_state.randint(self.n_features)
   698       519        437.0      0.8      0.0                  if terminal == self.n_features:
   699       106        261.0      2.5      0.0                      terminal = random_state.uniform(*self.const_range)
   700       106         73.0      0.7      0.0                      if self.const_range is None:
   701                                                                   # We should never get here
   702                                                                   raise ValueError('A constant was produced with '
   703                                                                                    'const_range=None.')
   704       519        428.0      0.8      0.0                  program[node] = terminal
   705                                           
   706      1165       2519.0      2.2      0.0          return program, list(mutate)
```

**With shallow copy:**
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   656                                               @profile
   657                                               def point_mutation(self, random_state):
   658                                                   """Perform the point mutation operation on the program.
   659                                           
   660                                                   Point mutation selects random nodes from the embedded program to be
   661                                                   replaced. Terminals are replaced by other terminals and functions are
   662                                                   replaced by other functions that require the same number of arguments
   663                                                   as the original node. The resulting tree forms an offspring.
   664                                           
   665                                                   Parameters
   666                                                   ----------
   667                                                   random_state : RandomState instance
   668                                                       The random number generator.
   669                                           
   670                                                   Returns
   671                                                   -------
   672                                                   program : list
   673                                                       The flattened tree representation of the program.
   674                                           
   675                                                   """
   676      1165       5726.0      4.9     12.0          program = copy(self.program)
   677                                           
   678                                                   # Get the nodes to modify
   679      1165       1401.0      1.2      2.9          mutate = np.where([True if (random_state.uniform() <
   680                                                                               self.p_point_replace)
   681                                                                      else False
   682      1165      28238.0     24.2     59.1                             for _ in range(len(program))])[0]
   683                                           
   684      2097       3342.0      1.6      7.0          for node in mutate:
   685       932        930.0      1.0      1.9              if isinstance(program[node], _Function):
   686       413        342.0      0.8      0.7                  arity = program[node].arity
   687                                                           # Find a valid replacement with same arity
   688       413        418.0      1.0      0.9                  replacement = len(self.arities[arity])
   689       413       1284.0      3.1      2.7                  replacement = random_state.randint(replacement)
   690       413        337.0      0.8      0.7                  replacement = self.arities[arity][replacement]
   691       413        316.0      0.8      0.7                  program[node] = replacement
   692                                                       else:
   693                                                           # We've got a terminal, add a const or variable
   694       519        375.0      0.7      0.8                  if self.const_range is not None:
   695       519       1581.0      3.0      3.3                      terminal = random_state.randint(self.n_features + 1)
   696                                                           else:
   697                                                               terminal = random_state.randint(self.n_features)
   698       519        414.0      0.8      0.9                  if terminal == self.n_features:
   699       106        193.0      1.8      0.4                      terminal = random_state.uniform(*self.const_range) # program[node] + 
   700       106         99.0      0.9      0.2                      if self.const_range is None:
   701                                                                   # We should never get here
   702                                                                   raise ValueError('A constant was produced with '
   703                                                                                    'const_range=None.')
   704       519        362.0      0.7      0.8                  program[node] = terminal
   705                                           
   706      1165       2446.0      2.1      5.1          return program, list(mutate)
```

**Comment by trevorstephens** [comment]
Interesting. The back-end has undergone a lot of changes over the years. I believe this was necessary in the original implementation, but how the mathematical functions were represented, and other program objects were stored was changed many times since then. Will investigate, thanks for the report.

**Comment by trevorstephens** [comment]
I just did a test with very high change of point mutation and high amount of it. The results were completely different when removing the deepcopy because I believe the parent will have their DNA changed with their children and reproduction will refer back to the parents, as well as their children. Happy to be proven wrong but this seems to break the algorithm. 

In the default settings, reproduction and point mutation aren't big players, wonder how much this *really* affects people?

That said it does appear to be a bit of a computational sponge. Might be a better way to go about it. Perhaps an array slice with substitution that way would be more efficient.

**Comment by wulfihm** [comment]
You need to replace the deepcopy with a shallow copy, meaning doing copy() instead of deepcopy().
If I do exactly that the results are the same.

**Comment by trevorstephens** [comment]
Yeah I think something went wrong in my test, very sorry about that @wulfihm !

I just tried again and indeed, results are the same. I made a very reproduction and point mutation heavy toy example and run times went down from 75 seconds to 4 seconds. Pretty huge difference. I doubt this matches any real world use but it is still an improvement in efficiency which is always welcome. Thanks again for the report. Should have this merged shortly in #139 .
